### PR TITLE
feature(dspy): generate a graphical representation of modules

### PR DESCRIPTION
**Pull request by fsndzomga**
The ModuleGraph class is a utility for generating visual representations of Python modules and their components using the graphviz library. It enables users to visualize the structure and interactions of modules, making it easier to understand complex systems.

### Key Features:

**Dynamic Graph Generation:** The ModuleGraph class dynamically generates a graph representation of Python modules, including their submodules and components.

**Dependency Visualization:** It visualizes dependencies between modules, submodules, and components, helping users understand how different parts of the system interact with each other.

**Graphviz Integration:** Uses the graphviz library to create and render graphs in various formats (e.g., PNG), allowing for customization and easy sharing of visualizations.

**Error Handling:** Provides informative error messages if the graphviz library is not installed, guiding users on how to install the required dependency (graphviz) via pip.

**Modular Design:** The class is designed with modularity in mind, making it easy to extend and customize for different use cases and module structures.

### Usage Example:

```python

import dspy
import os
from dotenv import load_dotenv

load_dotenv()

# Configuration of dspy models
llm = dspy.OpenAI(
    model='gpt-3.5-turbo',
    api_key=os.environ['OPENAI_API_KEY'],
    max_tokens=100
)

colbertv2_wiki = dspy.ColBERTv2(url='http://20.102.90.50:2017/wiki17_abstracts')

dspy.settings.configure(lm=llm, rm=colbertv2_wiki)

class GenerateAnswer(dspy.Signature):
  "Answer with long and detailed answers"
  context = dspy.InputField(desc="may content relevant facts")
  question = dspy.InputField()
  answer = dspy.OutputField(desc="often between 10 and 50 words")

class RAG(dspy.Module):
  def __init__(self, num_passages=3):
    super().__init__()
    self.retrieve = dspy.Retrieve(k=num_passages)
    self.generate_answer = dspy.ChainOfThought(GenerateAnswer)

  def forward(self, question):
    context = self.retrieve(question).passages
    prediction = self.generate_answer(context=context, question=question)
    return dspy.Prediction(context=context, answer=prediction.answer)

rag_system = RAG()
graph = ModuleGraph("RAG", rag_system)

graph.render_graph()

```
Here is the graph generate in this case:

Screenshot 2024-05-05 at 05 38 04

### CODE REVIEW

File: dspy/experimental/module_graph.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                Run 'pip install graphviz'""")
+
+        self.graph = graphviz.Digraph(format='png')
+        self.nodes = set()
+        self.module_name = module_name
+        self.module = module
+        self.inspect_settings(dspy.settings)
+        self.add_module(self.module_name, self.module)
+
+    def inspect_settings(self, settings):
```
Comment (arnavsinghvi11) [inline dspy/experimental/module_graph.py:24]: curious how setting a LM/RM in using `with dspy.settings.context(...):` ([reference example](https://github.com/stanfordnlp/dspy/blob/abf98a1491ee26a3bfc55c9a8abc340783e55dce/dspy/predict/langchain.py#L74)) of setting LM/RM to impact only the corresponding code block) -  impacts this functionality. is it better to instead have user-configurable parameters when initializing ModuleGraph?

File: dspy/experimental/module_graph.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        if filename is None:
+            filename = self.module_name
+        self.graph.render(filename)
+
+
+
+# Example usage of the ModuleGraph class:
+
+# class GenerateAnswer(dspy.Signature):
+#   "Answer with long and detailled answers"
```
Comment (arnavsinghvi11) [inline dspy/experimental/module_graph.py:107]: typo - "detailed"

File: dspy/experimental/module_graph.py (hunk 1 above)
Comment (arnavsinghvi11) [reply dspy/experimental/module_graph.py:24]: and what happens if these are not set?

File: dspy/experimental/module_graph.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#     super().__init__()
+#     self.retrieve = dspy.Retrieve(k=num_passages)
+#     self.generate_answer = dspy.ChainOfThought(GenerateAnswer)
+
+#   def forward(self, question):
+#     context = self.retrieve(question).passages
+#     prediction = self.generate_answer(context=context, question=question)
+#     return dspy.Prediction(context=context, answer=prediction.answer)
+
+# rag_system = RAG()
```
Comment (arnavsinghvi11) [inline dspy/experimental/module_graph.py:123]: would be good to add the configurations of the LM/RM in DSPy for accurate reproducibility

ex:
```dspy.settings.configure(lm=turbo, rm=colbertv2_wiki17_abstracts)`

**Comment by arnavsinghvi11** [comment]
Thanks @fsndzomga, this looks great! Left some minor comments and is ready to merge after that. 

It would be great to have documentation for it as well, potentially detailing the motivations behind use cases and applications for graphical representations added to the [Community Examples](https://dspy-docs.vercel.app/docs/tutorials/examples) if you would like to!

**Comment by arnavsinghvi11** [comment]
@fsndzomga just bumping this!

File: dspy/experimental/module_graph.py (hunk 1 above)
Comment (fsndzomga) [reply dspy/experimental/module_graph.py:24]: the current implementation only handle default LLMs. Can add LLMs in dspy.context later.

If nothing is set in dspy.settings.configure, then nothing will appear in the graph as expected.

File: dspy/experimental/module_graph.py (hunk 2 above)
Comment (fsndzomga) [reply dspy/experimental/module_graph.py:107]: Thanks for catching that one. will definitely correct that

File: dspy/experimental/module_graph.py (hunk 3 above)
Comment (fsndzomga) [reply dspy/experimental/module_graph.py:123]: Sure, no problem.

**Comment by arnavsinghvi11** [comment]
Hi @fsndzomga , just bumping the comments again as the PR is not yet mergeable. would love to merge this once those are resolved!

**Comment by fsndzomga** [comment]
> Hi @fsndzomga , just bumping the comments again as the PR is not yet mergeable. would love to merge this once those are resolved!

Hi @arnavsinghvi11, sorry, this completely slipped my mind. I just shipped the first version. I will further enhance the capabilities later.

**Comment by arnavsinghvi11** [comment]
Thanks @fsndzomga ! Merging this, and feel free to open another PR with additional changes and documentation!

**Pull request merged**